### PR TITLE
azure: fix cleaning up resources

### PR DESCRIPTION
**Pull request by knrt10**
VM, nic and disk were not getting cleaned up when instance deletion request was made.

Fixes: #612

### CODE REVIEW

File: pkg/adaptor/cloud/azure/provider.go [hunk 1]
```diff
@@ -251,7 +252,10 @@ func (p *azureProvider) DeleteInstance(ctx context.Context, instanceID string) e
 		return fmt.Errorf("creating VM client: %w", err)
 	}
 
-	pollerResponse, err := vmClient.BeginDelete(ctx, p.serviceConfig.ResourceGroupName, instanceID, nil)
+	// instanceID in the form of /subscriptions/<subID>/resourceGroups/<resource_name>/providers/Microsoft.Compute/virtualMachines/<VM_Name>.
+	vmName := strings.SplitAfter(instanceID, "virtualMachines/")[1]
```
Comment (mkulke) [inline pkg/adaptor/cloud/azure/provider.go:256]: It feels a bit unfortunate that we have to mangle strings like this. If the azure api doesn't accept an id, but a name as an input for delete operations. we maybe should treat that name as an identifier for the azure code (afaik within a sub + resource-group a name would be a unique identifier). 

In any case I think we might want to be extra diligent with this operation, afaik delete operation will not return fail on non-existing operations (at least that's the behaviour for `az vm delete`. Some ideas:

- cover a failure of string extraction + return an error, so there's at least a log
- use a regex to extract it to be sure we are mangling a proper string:
`^/subscriptions/$sub/resourceGroups/$rg/providers/Microsoft.Compute/virtualMachines/(.*)$`

File: pkg/adaptor/cloud/azure/provider.go [hunk 2]
```diff
@@ -260,7 +264,63 @@ func (p *azureProvider) DeleteInstance(ctx context.Context, instanceID string) e
...
+
+	logger.Printf("deleted disk successfully: %s", diskName)
+
+	return nil
+}
+
+func (p *azureProvider) deleteNetworkInterface(ctx context.Context, nicName string) error {
+	nicClient, err := armnetwork.NewInterfacesClient(p.serviceConfig.SubscriptionId, p.azureClient, nil)
+	if err != nil {
+		return fmt.Errorf("creating network interfaces client: %w", err)
```
Comment (mkulke) [inline pkg/adaptor/cloud/azure/provider.go:310]:
```suggestion
		return fmt.Errorf("creating network interface client: %w", err)
```

File: pkg/adaptor/cloud/azure/provider.go [hunk 3]
```diff
@@ -260,7 +264,63 @@ func (p *azureProvider) DeleteInstance(ctx context.Context, instanceID string) e
...
+	}
+
+	pollerResponse, err := nicClient.BeginDelete(ctx, p.serviceConfig.ResourceGroupName, nicName, nil)
+	if err != nil {
+		return fmt.Errorf("beginning deletion of network interface: %w", err)
```
Comment (mkulke) [inline pkg/adaptor/cloud/azure/provider.go:315]:
```suggestion
		return fmt.Errorf("beginning network interface deletion: %w", err)
```

File: pkg/adaptor/cloud/azure/provider.go [hunk 4]
```diff
@@ -260,7 +264,63 @@ func (p *azureProvider) DeleteInstance(ctx context.Context, instanceID string) e
...
+	}
+
+	_, err = pollerResponse.PollUntilDone(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("polling network interface deletion: %w", err)
```
Comment (mkulke) [inline pkg/adaptor/cloud/azure/provider.go:320]:
```suggestion
		return fmt.Errorf("waiting for network interface deletion: %w", err)
```

File: pkg/adaptor/cloud/azure/provider.go (hunk 1 above)
Comment (knrt10) [reply pkg/adaptor/cloud/azure/provider.go]: Make sense, updated

**Review by mkulke — APPROVED** [review]
thx. lgtm

**Review by bpradipt — APPROVED** [review]
/lgtm
Thanks @knrt10

**Pull request merged**